### PR TITLE
Adds system/environmental variable functions

### PR DESCRIPTION
**Pull request by Izaakwltn**
This adds system/environmental variable functions. 

`defenv`/`setenv` behavior is `Result` protected, with rules modeled after `defvar`/`setq`. 

Command line arguments are also `Result` protected.

### CODE REVIEW

File: library/system.lisp [hunk 1]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+  (define-instance (Signalable SysError)
+    (define (error (SysError msg obj))
+      (error (lisp String (msg obj)
+               (cl:format cl:nil "System Error:~%~%~a ~a" msg obj)))))
+
+  ;;
+  ;; Accessing Environment Variables
+  ;;
+  
+  (declare getenv (String -> (Result SysError String)))
```
Comment (eliaslfox) [inline library/system.lisp:88]: This should return optional

File: library/system.lisp [hunk 2]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+    "Gets the value of the environmental variable `var`, errors if `var` doesn't exist."
+    (lisp (Result SysError String) (var)
+               (cl:let ((env (uiop:getenvp var)))
+                 (cl:if env
+                        (Ok env)
+                        (Err (SysError "Environment variable not found:" var))))))
+
+  
+  (declare setenv (String -> String -> (Result SysError Unit)))
+  (define (setenv var val)
```
Comment (eliaslfox) [inline library/system.lisp:99]: don't define this function

File: library/system.lisp [hunk 3]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+    (match (getenv var)
+      ((Err x)
+       (Err x))
+      ((Ok _)
+       (lisp Integer (var val)
+           (cl:setf (uiop:getenv var) val))
+       (Ok Unit))))
+
+  (declare defenv (String -> String -> (Result SysError Unit)))
+  (define (defenv var val)
```
Comment (eliaslfox) [inline library/system.lisp:110]: call this `setenv`

File: library/system.lisp [hunk 4]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+       (Err (SysError "Environment variable already exists:" var)))))
+
+  ;;
+  ;; Typical Environment/System variables
+  ;;
+  
+  (declare architecture (String))
+  (define architecture
+    "The system's architecture (stored at compile time)."
+    (lisp String ()
```
Comment (eliaslfox) [inline library/system.lisp:127]: This and similar definitions should be functions.

File: library/system.lisp [hunk 5]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+
+  (declare lisp-version (String))
+  (define lisp-version
+    "The lisp implementation version (stored at compile time)."
+    (lisp String ()
+      (uiop:lisp-version-string)))
+
+  (declare lisp-impl-directory (String))
+  (define lisp-impl-directory
+    "The lisp implementation's directory (stored at compile time)."
```
Comment (eliaslfox) [inline library/system.lisp:156]: Is this useful?

File: library/system.lisp [hunk 6]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+    (lisp String ()
+      (cl:namestring (uiop:lisp-implementation-directory))))
+
+  (declare configuration-pathnames (Unit -> (List String)))
+  (define (configuration-pathnames)
+    "Returns the list of directories which store default user configuration."
```
Comment (eliaslfox) [inline library/system.lisp:162]: Is this useful?

File: library/system.lisp [hunk 7]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+    (lisp (List String) ()
+      (uiop:system-config-pathnames)))
+
+  (declare features (Unit -> (List String)))
+  (define (features)
+    "Returns a list of active features, from `cl:*features*`."
+    (lisp (list String) ()
+      (cl:mapcar #'cl:string cl:*features*)))
```
Comment (eliaslfox) [inline library/system.lisp:170]: Maybe `(cl:mapcar #'cl:symbol-name ...)`

File: library/system.lisp [hunk 8]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+  ;; Command line arguments
+  ;;
+  
+  (declare cmd-args ((Result SysError (List String))))
+  (define cmd-args
+    "The current command line arguments (stored at compile time)."
+    (lisp (Result SysError (List String)) ()
+      (cl:let ((cla (uiop:command-line-arguments)))
+        (cl:if (cl:null cla)
+               (Err (SysError "No command line arguments found." ""))
```
Comment (eliaslfox) [inline library/system.lisp:191]: This should just return an empty list instead of erroring.

File: library/system.lisp [hunk 9]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
+               (Ok cla)))))
+
+  (declare argv0 ((Result SysError String)))
+  (define argv0
+    "The first command line argument (stored at compile time)."
```
Comment (eliaslfox) [inline library/system.lisp:196]: This should be a function.

File: library/system.lisp [hunk 10]
```diff
@@ -45,5 +62,142 @@ While the result will always contain microseconds, some implementations may retu
...
 
+;;;
+;;; Gathering System information
+;;;
+
+(coalton-toplevel
+
+  (define-type SysError
+    "An error message for system errors."
+    (SysError String String))
```
Comment (eliaslfox) [inline library/system.lisp:73]: Why two strings?

Comment (Izaakwltn) [reply library/system.lisp]: Actually since I'm using `Optional` instead of result I just got rid of `SysError` as a whole.

**Review by eliaslfox — APPROVED** [review]
nice work

**Pull request merged**